pallets/phala/src/fat.rs

```diff
@@ -377,45 +428,20 @@ pub mod pallet {
 			}
 			Ok(())
 		}
+	}
 
-		pub fn on_worker_contract_message_received(
-			message: DecodedMessage<WorkerContractReport>,
-		) -> DispatchResult {
-			let _worker_pubkey = match &message.sender {
-				MessageOrigin::Worker(worker_pubkey) => worker_pubkey,
-				_ => return Err(Error::<T>::InvalidSender.into()),
-			};
-			match message.payload {
-				WorkerContractReport::ContractInstantiated {
-					id,
-					cluster_id,
-					deployer,
-					pubkey: _,
-				} => {
-					let contracts = ClusterContracts::<T>::get(&cluster_id);
-					if !contracts.contains(&id) {
-						ClusterContracts::<T>::append(&cluster_id, &id);
-					}
-					Self::deposit_event(Event::Instantiated {
-						contract: id,
-						cluster: cluster_id,
-						deployer,
-					});
-				}
-				WorkerContractReport::ContractInstantiationFailed {
-					id,
-					cluster_id,
-					deployer,
-				} => {
-					Self::deposit_event(Event::InstantiationFailed {
-						contract: id,
-						cluster: cluster_id,
-						deployer,
-					});
-					// TODO.shelven: some cleanup?
-				}
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(_now: BlockNumberFor<T>) -> Weight {
+			if let Some(next_code) = NextPinkSystemCode::<T>::take() {
+				let hash: H256 = crate::hashing::blake2_256(&next_code).into();
+				PinkSystemCodeHash::<T>::put(hash);
+				PinkSystemCode::<T>::mutate(|(ver, code)| {
```

Comment:
Do we want to push a message to workers to inform them to upgrade to the new system code?

Reply:
It is automatically used by new cluster deployment.
For already deployed clusters, we haven't implemented any upgradable contract yet. Do you think we should implement it in the first version?

Reply:
Not a big problem. We can add it later